fastcrypto/src/hash.rs

```diff
@@ -2,126 +2,151 @@
     fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
+        self.digest.as_ref()
     }
 }
 
-/// This trait is implemented by all messages that can be hashed.
-pub trait Hashable<DigestLength: ArrayLength<u8> + 'static> {
-    fn digest<H: HashFunction<DigestLength>>(self) -> Digest<DigestLength>;
-}
-
-impl<DigestLength: ArrayLength<u8> + 'static> Hashable<DigestLength> for &[u8] {
-    /// Hash this data using the given hash function.
-    fn digest<H: HashFunction<DigestLength>>(self) -> Digest<DigestLength> {
-        H::digest(self)
+impl<const DIGEST_LEN: usize> From<Digest<DIGEST_LEN>> for [u8; DIGEST_LEN] {
+    fn from(digest: Digest<DIGEST_LEN>) -> Self {
+        digest.digest
     }
 }
 
 /// Trait implemented by hash functions providing a output of fixed length
-pub trait HashFunction<DigestLength: ArrayLength<u8>>: OutputSizeUser + Sized + Default {
+pub trait HashFunction<const DIGEST_LENGTH: usize>: Default {
+    /// The length of this hash functions digests in bytes.
+    const OUTPUT_SIZE: usize = DIGEST_LENGTH;
+
+    /// Create a new hash function of the given type
+    fn new() -> Self {
+        Self::default()
+    }
+
     /// Process the given data, and update the internal of the hash function.
-    fn update(&mut self, data: &[u8]);
+    fn update<Data: AsRef<[u8]>>(&mut self, data: Data);
 
     /// Retrieve result and consume hash function.
-    fn finalize(self) -> Digest<DigestLength>;
+    fn finalize(self) -> Digest<DIGEST_LENGTH>;
 
-    fn digest(data: &[u8]) -> Digest<DigestLength> {
+    /// Compute the digest of the given data and consume the hash function.
+    fn digest<Data: AsRef<[u8]>>(data: Data) -> Digest<DIGEST_LENGTH> {
         let mut h = Self::default();
         h.update(data);
         h.finalize()
     }
+
+    /// Compute a single digest from all slices in the iterator in order and consume the hash function.
+    fn digest_iterator<K: AsRef<[u8]>, I: Iterator<Item = K>>(iter: I) -> Digest<DIGEST_LENGTH> {
+        let mut h = Self::default();
+        iter.into_iter().for_each(|chunk| h.update(chunk.as_ref()));
+        h.finalize()
+    }
 }
 
-#[derive(Default)]
-pub struct HashFunctionWrapper<Variant: digest::Digest + 'static>(Variant);
+/// This trait is implemented by all messages that can be hashed.
+pub trait Hash<const DIGEST_LEN: usize> {
+    /// The type of the digest when this is hashed.
+    type TypedDigest: Into<Digest<DIGEST_LEN>> + Eq + std::hash::Hash + Copy;
 
-impl<Variant: digest::Digest + 'static> OutputSizeUser for HashFunctionWrapper<Variant> {
-    type OutputSize = Variant::OutputSize;
+    fn digest(&self) -> Self::TypedDigest;
 }
 
-impl<Variant: digest::Digest + 'static + Default> HashFunction<Variant::OutputSize>
-    for HashFunctionWrapper<Variant>
+/// This wraps a `digest::Digest` as a `fastcrypto::hash::HashFunction`.
+#[derive(Default)]
+pub struct HashFunctionWrapper<Variant: digest::Digest + 'static, const DIGEST_LEN: usize>(Variant);
+
```

Comment:
We need both the `Variant`type and the `DIGEST_LEN`constant because the digest length in the underlying crates is defined as a type but we use a constant. This struct should only be used internally though.